.changeset/sixty-coins-worry.md

```diff
@@ -0,0 +1,29 @@
+---
+'astro': minor
+---
+
+Adds support for passing values other than `"production"` or `"development"` to the `--mode` flag, e.g. `"staging"`, `"testing"`, or any value. This allows to change the value of `import.meta.env.MODE` or the loaded `.env` file, and take advantage of Vite's [mode](https://vite.dev/guide/env-and-mode#modes) feature.
+
+Note that changing the `mode` does not change the kind of code transform handled by Vite and Astro:
+
+- In `astro dev`, Astro will transform code with debug information.
+- In `astro build`, Astro will transform code with the most optimized output and removes debug information.
+- In `astro build --dev` (new flag), Astro will transform code with debug information like in `astro dev`. 
+
+This enables various usecases like:
+
+```bash
+# Run the dev server connected to a "staging" API
+astro dev --mode staging
+
+# Build a site that connects to a "staging" API
+astro build --mode staging
+
+# Build a site that connects to a "production" API with additional debug information
+astro build --dev
```

Comment:
Why the `--dev` flag vs `--mode development` ?

Reply:
Someone might want `--mode testing` perhaps, but with development behaviour, so we need another way to hint that and `--dev` fills that gap. `--mode development` would only work if the user chooses it directly, but the goal of the PR is to allow different mode names to be used and picking between prod/dev behaviours.

Reply:
In this specific example, it also assumes that you have `.env.production` that contains some env vars to a production `API_URL` perhaps. So this config allows you to keep using that production `API_URL`, but with dev debug information.

Reply:
I'm confused by what "dev behavior" is. If this is just about adding debug symbols, can it be named more specifically?

Also, are these different features? If so maybe separate PRs is better.

Reply:
Example of "dev behaviour" and "prod behaviour":

1. `react` exports a `react.development.js` and `react.production.js` bundles. The former has better debugging information, full error names, etc. The latter has shorten error codes and minified code.
2. `svelte`'s compiler has a `dev` option, which can output elaborate code with better error messages and guardrails. And if it's disabled, those will be stripped away.
3. `vue` Vite plugin checks against dev or prod to decide injecting a script to connect to the browser devtools. This allows you to inspect the Vue component tree and debug things. If prod, it'll not be injected and you can't debug it.

So there needs to be a way to toggle between these two. `mode` is no longer the way to tell apart if we want the "dev behaviour" or "prod behaviour". The new way is documented above at `Within Astro, how do we actually tell in we're in dev or prod state now?`

`--dev` is needed to take part of this toggling behaviour.

Reply:
For the most part, Vite infers this via `process.env.NODE_ENV` and will handle the rest. We set it here: https://github.com/withastro/astro/blob/c1cebb4b2dc7fedc291d92503a750d2871d2e4d3/packages/astro/src/core/build/index.ts#L62

- `react` uses the `"development"` and `"production"` exports condition, which Vite will handle by inferring `process.env.NODE_ENV`
- `svelte` and `vue` derives this from the Vite plugin's `resolvedConfig.isProduction` variable, which in turn is also derived from `process.env.NODE_ENV` too.

So we only need to set `process.env.NODE_ENV` before Vite starts up and our job is done 😄

Reply:
Ok, in that case could it be named `--node-env-dev`? Also, is this enough of an improvement over `NODE_ENV=development astro build` to justify?

I don't mind having this feature, I just want to avoid unmet expectations because it's vague what "dev" means.

Reply:
`--node-env-dev` and `NODE_ENV=development` feels like exposing implementation details to users 🤔 It's also a bit long and clunky to type in a script command. I definitely get that `build --dev` can be confusing, e.g. "is it watch mode, hmr in builds, etc", maybe something like `--dev-output`?

The flag would also exists as an option for the `build()` JS API, so a name needs to be mapped there too.

Reply:
I believe we should remove the concept of **dev**. Other ecosystems, for example Rust, use the concept of **debug**. So, to make an example from the original proposal:
```diff
-astro build --dev
+astro build --debug
```

The concept of "dev"/"development" is usually linked to something that is meant for **local** development. However, if a user decides to do something like `astro build --dev`, it's more likely they want to *debug* their application in a production-like environment, but with debug symbols enabled.

I propose to expose the term "debug" **to the users** and avoid to use things like `--node-env-dev` because it's exposing implementation details to the users (I agree with @bluwy here).

Reply:
I kinda like `--debug`. My first instinct with that is that it could imply a debug log level, but we have that named `--verbose` now, so maybe it's not too bad.

Reply:
I'll change it as `--dev-output` for now then and we could discuss it later again if needed. WIll also create docs and see what the docs team thinks of this.